### DISCUSSION TITLE
How to store credentials in relational database (postgres / mariadb / mysql)

### DISCUSSION THREAD

**Discussion by uvulpos** [go-webauthn/webauthn#219] [Q&A] [answered]
As a Software Engineer, I want to persist the created credential into my database, but currently I don't know what I have to persist exactly into the database and how to create the schema. Which columns are important? Is it also possible to prevent users from using the same YubiKey on different accounts, or is the browser smart enough to detect that?

**Comment by james-d-elliott** [comment, marked as answer]
I'll try to consolidate most of this information into a FAQ at some stage as a number of people have been asking.

1. All of the data in the webauthn.Session struct returned from various functions must be in the same state between the Begin and Finish/Validate function pairs. i.e. if you use BeginLogin the webauthn.Session it returns should be provided to the FinishLogin function with the same field values. How you achieve this is up to you.
2. The `WebAuthnID` implementation for the `User` interface should always return the same value. Generally speaking this should be an opaque value. This value is what I store in the `webauthn_users` table with the `userid` column. It should be noted there is a convenience option for this here which adapts the way this field is encoded [here](https://github.com/go-webauthn/webauthn/blob/master/webauthn/types.go#L51C2-L51C22) (the `EncodeUserIDAsString` field of the `Config` struct in the `github.com/go-webauthn/webauthn/webauthn` module).
3. All of the values within the webauthn.Credentials struct should realistically be saved and some should be updated in storage on each authentication. Roughly:
    1. The `ID` field which I store in the `kid` column.
    2. The `PublicKey` field which I store in the `public_key` column.
    2. The `AttestationType` field which I store in the `attestation_type` column.
    3. The `Transport` field which I store in the `transport` column.
    3. The `Authenticator.AAGUID` field which I store in the `aaguid` column.
    4. The `Authenticator.SignCount` field which I store in the `sign_count` column, this should also be updated on every successful FinishLogin/ValidateLogin.
    5. The `Authenticator.CloneWraning` field which should be updated on every sucessful FinishLogin/ValidateLogin.
    6. The `Flags.UserPresent` flag which I store in the `present` column.
    7. The `Flags.UserVerified` flag which I store in the `verified` column.
    8. The `Flags.BackupEligible` flag which I store in the `backup_eligible` column.
    9. The `Flags.BackupState` flag which should also be updated on every authentication if Flags.BackupEligible is true which I store in the `backup_state` column.
    
I also personally store the results from the `credProps` extensions specifically the `rk` value is stored in the `discoverable` column.

I have included a rough schema I've used for PostgreSQL below which covers 2 and 3. I personally encrypt the `public_key` locally in my application so that the value itself can't be tampered with unless you have the encryption key (that's optional). 

```sql
CREATE TABLE IF NOT EXISTS webauthn_credentials (
    id SERIAL CONSTRAINT webauthn_credentials_pkey PRIMARY KEY,
    created_at TIMESTAMP WITH TIME ZONE NOT NULL DEFAULT CURRENT_TIMESTAMP,
    last_used_at TIMESTAMP WITH TIME ZONE NULL DEFAULT NULL,
    rpid VARCHAR(512) NOT NULL,
    username VARCHAR(100) NOT NULL,
    description VARCHAR(64) NOT NULL,
    kid VARCHAR(512) NOT NULL,
    aaguid CHAR(36) NULL,
    attestation_type VARCHAR(32),
    attachment VARCHAR(64) NOT NULL,
    transport VARCHAR(64) DEFAULT '',
    sign_count INTEGER DEFAULT 0,
    clone_warning BOOLEAN NOT NULL DEFAULT FALSE,
    discoverable BOOLEAN NOT NULL,
    present BOOLEAN NOT NULL DEFAULT FALSE,
    verified BOOLEAN NOT NULL DEFAULT FALSE,
    backup_eligible BOOLEAN NOT NULL DEFAULT FALSE,
    backup_state BOOLEAN NOT NULL DEFAULT FALSE,
    public_key BYTEA NOT NULL
);

CREATE UNIQUE INDEX webauthn_credentials_kid_key ON webauthn_credentials (kid);
CREATE UNIQUE INDEX webauthn_credentials_lookup_key ON webauthn_credentials (rpid, username, description);

CREATE TABLE IF NOT EXISTS webauthn_users (
    id SERIAL CONSTRAINT webauthn_users_pkey PRIMARY KEY,
    rpid VARCHAR(512) NOT NULL,
    username VARCHAR(100) NOT NULL,
    userid CHAR(64) NOT NULL
);

CREATE UNIQUE INDEX webauthn_users_lookup_key ON webauthn_users (rpid, username);
```

**Reply by james-d-elliott** [reply]
Example that can be used for advanced post-registration metadata validation:

```sql
CREATE TABLE IF NOT EXISTS webauthn_credentials (
    id SERIAL CONSTRAINT webauthn_credentials_pkey PRIMARY KEY,
    created_at TIMESTAMP WITH TIME ZONE NOT NULL DEFAULT CURRENT_TIMESTAMP,
    last_used_at TIMESTAMP WITH TIME ZONE NULL DEFAULT NULL,
    rpid VARCHAR(512) NOT NULL,
    username VARCHAR(100) NOT NULL,
    description VARCHAR(64) NOT NULL,
    kid VARCHAR(512) NOT NULL,
    aaguid CHAR(36) NULL,
    attestation_type VARCHAR(32),
    attachment VARCHAR(64) NOT NULL,
    transport VARCHAR(64) DEFAULT '',
    sign_count INTEGER DEFAULT 0,
    clone_warning BOOLEAN NOT NULL DEFAULT FALSE,
    legacy BOOLEAN NOT NULL DEFAULT FALSE,
    discoverable BOOLEAN NOT NULL,
    present BOOLEAN NOT NULL DEFAULT FALSE,
    verified BOOLEAN NOT NULL DEFAULT FALSE,
    backup_eligible BOOLEAN NOT NULL DEFAULT FALSE,
    backup_state BOOLEAN NOT NULL DEFAULT FALSE,
    attestation BYTEA NULL DEFAULT NULL,
    public_key BYTEA NOT NULL
);

CREATE UNIQUE INDEX webauthn_credentials_kid_key ON webauthn_credentials (kid);
CREATE UNIQUE INDEX webauthn_credentials_lookup_key ON webauthn_credentials (rpid, username, description);

CREATE TABLE IF NOT EXISTS webauthn_users (
    id SERIAL CONSTRAINT webauthn_users_pkey PRIMARY KEY,
    rpid VARCHAR(512) NOT NULL,
    username VARCHAR(100) NOT NULL,
    userid CHAR(64) NOT NULL
);

CREATE UNIQUE INDEX webauthn_users_lookup_key ON webauthn_users (rpid, username);
```

**Reply by ayushs-2k4** [reply]
> Example that can be used for advanced post-registration metadata validation:
> 
> ```sql
> CREATE TABLE IF NOT EXISTS webauthn_credentials (
>     id SERIAL CONSTRAINT webauthn_credentials_pkey PRIMARY KEY,
>     created_at TIMESTAMP WITH TIME ZONE NOT NULL DEFAULT CURRENT_TIMESTAMP,
>     last_used_at TIMESTAMP WITH TIME ZONE NULL DEFAULT NULL,
>     rpid VARCHAR(512) NOT NULL,
>     username VARCHAR(100) NOT NULL,
>     description VARCHAR(64) NOT NULL,
>     kid VARCHAR(512) NOT NULL,
>     aaguid CHAR(36) NULL,
>     attestation_type VARCHAR(32),
>     attachment VARCHAR(64) NOT NULL,
>     transport VARCHAR(64) DEFAULT '',
>     sign_count INTEGER DEFAULT 0,
>     clone_warning BOOLEAN NOT NULL DEFAULT FALSE,
>     legacy BOOLEAN NOT NULL DEFAULT FALSE,
>     discoverable BOOLEAN NOT NULL,
>     present BOOLEAN NOT NULL DEFAULT FALSE,
>     verified BOOLEAN NOT NULL DEFAULT FALSE,
>     backup_eligible BOOLEAN NOT NULL DEFAULT FALSE,
>     backup_state BOOLEAN NOT NULL DEFAULT FALSE,
>     attestation BYTEA NULL DEFAULT NULL,
>     public_key BYTEA NOT NULL
> );
> 
> CREATE UNIQUE INDEX webauthn_credentials_kid_key ON webauthn_credentials (kid);
> CREATE UNIQUE INDEX webauthn_credentials_lookup_key ON webauthn_credentials (rpid, username, description);
> 
> CREATE TABLE IF NOT EXISTS webauthn_users (
>     id SERIAL CONSTRAINT webauthn_users_pkey PRIMARY KEY,
>     rpid VARCHAR(512) NOT NULL,
>     username VARCHAR(100) NOT NULL,
>     userid CHAR(64) NOT NULL
> );
> 
> CREATE UNIQUE INDEX webauthn_users_lookup_key ON webauthn_users (rpid, username);
> ```

I am still not able to implement this is postgres. I am retrieving values from DB but then my MacBook is not recognizing that a passkey is saved for "localhost", I tried to debug and I think it is due to Attestation props in red being null or 0, though I may be wrong. can you help me? like can you provide a code example how you implemented it using postgreSQL/ MySQL?

**Reply by nannany** [reply]
I am not quite sure what username and description refer to. Also, what is the index that webauthn_credentials_lookup_key is set for?

**Reply by james-d-elliott** [reply]
Username is business logic linking the to the user and description is a friendly name for the users to describe that credential, the lookup key may be used to modify a credential in various ways in certain scenarios.

**Comment by ptman** [comment]
I seem to have started using the library before `BackupEligible` was exposed. 0.11.0 now fails if you don't store and restore the value.

**Reply by jgimenez** [reply]
Same happened to us. Any ideas on how to deal with this? I'm a bit surprised the spec requires that the `BackupEligible` is sent and validated, if the RP is supposed to store it (why resend it then?) and if it's never supposed to change (if the crypto is sound, why check this one additional bit?).

I was thinking maybe we could make this validation optional? I would be happy to provide a PR.

**Reply by james-d-elliott** [reply]
I think you'd have to ask the spec authors on why it's not meant to change. I suspect the change in this flag would represent a flawed authenticator implementation but there is likely an additional security element associated.

I'd recommend to handle this that an additional value is stored to indicate the compatibility level of a credential and values are patched to reflect this. I've included an updated schema that I'm using which includes the attestation data as well as a legacy flag which could be leveraged to determine if the BE/BS flags (and legacy flag) are updated upon a login. I believe this would be possible using current methods if you parse and validate separately. If this wasn't the case then maybe we introduce tooling to allow for that.

For example adding an interface that indicates the BE/BS flags may not have been recorded at registration time:

```go
type BackupFlagCredential interface {
    GetBackupFlagsRegistered() bool
    SetBackupFlagsRegistered()
}
```

**Reply by jgimenez** [reply]
I tracked down this conversation that essentially says it's to detect a flawed authenticator implementation, as you suggested: https://github.com/w3c/webauthn/issues/1791

Therefore, I think it would be safe to:
 * Allow the user of the library to disable checking: trivial implementation, but not so nice.
 * Conditionally check the flag if stored during registration: I understand that's what you suggest. That would be spec compliant because the spec recommends but does not mandate the storage of this flag
 * Trust on first check: since the BE flag is not supposed to change, we could skip validation if we don't know the value but store it for next times.

> I believe this would be possible using current methods if you parse and validate separately. If this wasn't the case then maybe we introduce tooling to allow for that.

I think at the moment it's not possible to do that conditionally, because the check is an equality test:

```go
if credential.Flags.BackupEligible != parsedResponse.Response.AuthenticatorData.Flags.HasBackupEligible() {
	return nil, protocol.ErrBadRequest.WithDetails("BackupEligible flag inconsistency detected during login validation")
}
```

so we could protect this check with something like:

```
if GetBackupFlagsRegistered() { ... }
```

**Reply by james-d-elliott** [reply]
Almost certain you can do something like this.

```go
	parsedResponse, err := protocol.ParseCredentialRequestResponse(response)
	if err != nil {
		panic(err)
	}
	
	// Match the credential
	cred, err := DomainLogicGetCredentialFromUserFunc(parsedResponse, user)
	if err != nil {
		panic(err)
	}

        if cred.Legacy {
           cred.Flags.BackupEligible = parsedResponse.Response.AuthenticatorData.Flags.HasBackupEligible()
	   cred.Flags.BackupState = parsedResponse.Response.AuthenticatorData.Flags.HasBackupState()
	   cred.Legacy = false
        }	
        
        credential, err := ValidateLogin(user, session, parsedResponse)
```

**Reply by james-d-elliott** [reply]
That example is not strictly correct since you'd have to pre-modify the values stored in the user and have the user implementation build the []Credential slice from the known information. Essentially they'd be domain logic conerns. But the functions available currently make the concept of fixing lack of stored flags possible.

Regarding implementation I think adding some form of tooling to support something like this to migrate to proper storage is fine, the idea of baking something into the library to arbitrarily disable this check would be much more palatable for me if there were cases where proper credential storage did not specifically handle some case.

Examples of this would be an authenticator that is known to behave incorrectly particularly if that authenticator was also a certified authenticator.

**Reply by jgimenez** [reply]
That makes a lot of sense, thanks for the help!

**Reply by sarvranjan** [reply]
This is very useful, thanks guys.
Meanwhile, how do you propose we handle the "legacy" authenticators for validation?
It's not possible to try passing both true and false to BE flag as the credential ID needs to be updated to try both of these.

**Comment by ptman** [comment]
0.17.0 breaking changes may require actions https://github.com/go-webauthn/webauthn/releases/tag/v0.17.0 . (not documented in https://github.com/go-webauthn/webauthn/blob/master/BREAKING.md )

**Reply by james-d-elliott** [reply]
Fixed in ec84b4e6e17b3aad74e386cb2d9ef677311bfce6, the place I'll be placing breaking changes is in the release notes, especially as we gear up for v1.
